### PR TITLE
[PATCH API-NEXT v3] api: classification: adds packet hashing feature

### PR DESCRIPTION
**Pull request by bala-manoharan**
V3 incorporates review comments from Maxim
V2 incorporates doxygen review comments from Bill

### CODE REVIEW

File: platform/linux-generic/odp_classification.c [hunk 1]
```diff
@@ -89,8 +102,22 @@ int odp_classification_init_global(void)
 		LOCK_INIT(&pmr->s.lock);
 	}
 
+	queue_grp_shm = odp_shm_reserve("shm_odp_cls_queue_grp_tbl",
+					sizeof(_cls_queue_grp_tbl_t),
+					sizeof(queue_entry_t *), 0);
+
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/odp_classification.c]: This area is allocated but never freed. It should be in odp_classification_term_global(). Running the classification validation test gives the error:
```
_ishm.c:1659:_odp_ishm_term_global():block 'shm_odp_cls_queue_grp_tbl' (file ) was never freed (cleaning up...).
```

File: include/odp/api/spec/classification.h [hunk 2]
```diff
@@ -164,9 +171,40 @@ typedef enum {
...
+			 * packet input hash calculation */
+			odp_pktin_hash_proto_t hash_proto;
+		};
+	};
+	/** Pool associated with CoS */
+	odp_pool_t pool;
+
+	/** Drop policy associated with CoS */
+	odp_cls_drop_t drop_policy;
 } odp_cls_cos_param_t;
```
Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/classification.h]: There are still doxygen issues with these API changes. Not sure why Travis isn't picking this up. 

```
bill@Ubuntu64:~/linaro/balacos$ make doxygen-doc
  DXGEN  doc/application-api-guide/Doxyfile
/home/bill/linaro/balacos/include/odp/api/spec/classification.h:202: warning: parameters of member odp_cls_cos_param::@3 are not (all) documented
  DXGEN  doc/helper-guide/Doxyfile
  DXGEN  doc/platform-api-guide/Doxyfile
/home/bill/linaro/balacos/include/odp/api/spec/classification.h:202: warning: parameters of member odp_cls_cos_param::@3 are not (all) documented
  DXGEN  doc/driver-api-guide/Doxyfile
bill@Ubuntu64:~/linaro/balacos$ 
```

File: include/odp/api/spec/classification.h [hunk 3]
```diff
@@ -164,9 +171,40 @@ typedef enum {
...
+	 * When hashing is enabled the queues are created by the implementation
+	 * and application need not configure any queue to the class of service.
+	 * When hashing is disabled application has to configure the queue to
+	 * the class of service.
+	 * Depening on the implementation this number might be rounded-off to
+	 * nearest supported value (e.g power of 2)
+	 */
+	uint32_t num_queue;
+
+	union {
```
Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/classification.h]: Doxygen wants a comment for this union. Every struct/union/var needs some doxygen comment. Suggestion:

/** Variant mappings for queue hash configuration */

File: platform/linux-generic/odp_classification.c (hunk 1 above)
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/odp_classification.c]: @muvarov Can Travis detect these sort of _ishm.c warning messages? These should always be flagged as changes shouldn't introduce potential memory leaks.

File: include/odp/api/spec/classification.h (hunk 2 above)
Comment (lumag) [inline include/odp/api/spec/classification.h]: @Bill-Fischofer-Linaro Maybe because doxygen from trusty isn't so strict?

Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/classification.h]: @lumag I suspect you're correct. It's Doxygen 1.8.13 that has gotten super-strict.

File: platform/linux-generic/odp_classification.c (hunk 1 above)
Comment (bala-manoharan) [inline platform/linux-generic/odp_classification.c]: This has been fixed in the V2

**Comment by Bill-Fischofer-Linaro** [comment]
We still need an update to the classification section of the User Guide to cover these new changes. Bala, are you planning a separate PR for that?

**Comment by bala-manoharan** [comment]
Yes. I will create a separate PR for user-guide and validation suite

Regards,
Bala

File: platform/linux-generic/include/protocols/thash.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	union {
+		struct {
+			uint16_t sport;
+			uint16_t dport;
+		};
+		uint32_t sctp_tag;
+	};
+} thash_ipv6_tuple_t;
+
+#define ODP_THASH_V4_L4_LEN (sizeof(thash_ipv4_tuple_t) / 4)
```
Comment (bala-manoharan) [inline platform/linux-generic/include/protocols/thash.h]: Okay. will change in next version

**Comment by muvarov** [comment]
Merged!

**Pull request closed without merging**